New exception on 1.3.0.CR2 with kafka and without reasteasy

**Describe the bug**

On a project using kafka, undertow and vertx web (and not reasteasy), when upgrading from `1.3.0.Alpha1` to `1.3.0.CR2`, in jar mode, this exception is now popping (no problem before):
```
Exception in thread "main" java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:248)
        at io.quarkus.runtime.Application.start(Application.java:90)
        at io.quarkus.runtime.Application.run(Application.java:228)
        at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:41)
Caused by: java.lang.NoClassDefFoundError: org/jboss/resteasy/core/ResteasyContext
        at org.jboss.resteasy.context.ResteasyContextProvider.currentContext(ResteasyContextProvider.java:18)
        at io.smallrye.context.impl.CapturedContextState.<init>(CapturedContextState.java:21)
        at io.smallrye.context.SmallRyeContextManager.captureContext(SmallRyeContextManager.java:64)
        at io.smallrye.context.SmallRyeThreadContext.currentContextExecutor(SmallRyeThreadContext.java:208)
        at io.smallrye.context.propagators.rxjava2.ContextPropagatorOnFlowableAssemblyAction.apply(ContextPropagatorOnFlowableAssemblyAction.java:23)
        at io.smallrye.context.propagators.rxjava2.ContextPropagatorOnFlowableAssemblyAction.apply(ContextPropagatorOnFlowableAssemblyAction.java:11)
        at io.reactivex.plugins.RxJavaPlugins.apply(RxJavaPlugins.java:1268)
        at io.reactivex.plugins.RxJavaPlugins.onAssembly(RxJavaPlugins.java:1022)
        at io.vertx.reactivex.FlowableHelper.toFlowable(FlowableHelper.java:36)
        at io.vertx.reactivex.kafka.client.consumer.KafkaConsumer.toFlowable(KafkaConsumer.java:107)
        at io.smallrye.reactive.messaging.kafka.KafkaSource.<init>(KafkaSource.java:56)
        at io.smallrye.reactive.messaging.kafka.KafkaConnector.getPublisherBuilder(KafkaConnector.java:65)
        at io.smallrye.reactive.messaging.kafka.KafkaConnector_ClientProxy.getPublisherBuilder(KafkaConnector_ClientProxy.zig:203)
        at io.smallrye.reactive.messaging.impl.ConfiguredChannelFactory.createPublisherBuilder(ConfiguredChannelFactory.java:171)
        at io.smallrye.reactive.messaging.impl.ConfiguredChannelFactory.lambda$register$4(ConfiguredChannelFactory.java:149)
        at java.util.HashMap.forEach(HashMap.java:1289)
        at io.smallrye.reactive.messaging.impl.ConfiguredChannelFactory.register(ConfiguredChannelFactory.java:149)
        at io.smallrye.reactive.messaging.impl.ConfiguredChannelFactory.initialize(ConfiguredChannelFactory.java:123)
        at io.smallrye.reactive.messaging.impl.ConfiguredChannelFactory_ClientProxy.initialize(ConfiguredChannelFactory_ClientProxy.zig:195)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:647)
        at io.smallrye.reactive.messaging.extension.MediatorManager.initializeAndRun(MediatorManager.java:156)
        at io.smallrye.reactive.messaging.extension.MediatorManager_ClientProxy.initializeAndRun(MediatorManager_ClientProxy.zig:100)
        at io.quarkus.smallrye.reactivemessaging.runtime.SmallRyeReactiveMessagingLifecycle.onApplicationStart(SmallRyeReactiveMessagingLifecycle.java:20)
        at io.quarkus.smallrye.reactivemessaging.runtime.SmallRyeReactiveMessagingLifecycle_Observer_onApplicationStart_4e8937813d9e8faff65c3c07f88fa96615b70e70.notify(SmallRyeReactiveMessagingLifecycle_Observer_onApplicationStart_4e8937813d9e8faff65c3c07f88fa96615b70e70.zig:51)
        at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:282)
        at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:267)
        at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:69)
        at io.quarkus.arc.runtime.LifecycleEventRunner.fireStartupEvent(LifecycleEventRunner.java:23)
        at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:109)
        at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent39.deploy_0(LifecycleEventsBuildStep$startupEvent39.zig:58)
        at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent39.deploy(LifecycleEventsBuildStep$startupEvent39.zig:77)
        at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:210)
        ... 3 more
Caused by: java.lang.ClassNotFoundException: org.jboss.resteasy.core.ResteasyContext
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
        ... 37 more

```

CC @cescoffier

--- a/bom/runtime/pom.xml
+++ b/bom/runtime/pom.xml
@@ -2014,6 +2014,12 @@
                 <groupId>org.jboss.resteasy</groupId>
                 <artifactId>resteasy-rxjava2</artifactId>
                 <version>${resteasy.version}</version>
+                <exclusions>
+                    <exclusion>
+                        <groupId>org.jboss.resteasy</groupId>
+                        <artifactId>resteasy-context-propagation</artifactId>
+                    </exclusion>
+                </exclusions>
             </dependency>
             <dependency>
                 <groupId>org.jboss.resteasy</groupId>

--- a/build-parent/pom.xml
+++ b/build-parent/pom.xml
@@ -298,7 +298,7 @@
                                             <!-- We use our own impl here, including this one causes problems-->
                                             <exclude>org.jboss.resteasy:resteasy-context-propagation</exclude>
                                         </excludes>
-                                        <includes>a
+                                        <includes>
                                             <!-- this is for REST Assured -->
                                             <include>jakarta.xml.bind:jakarta.xml.bind-api:*:*:test</include>
                                         </includes>

--- a/build-parent/pom.xml
+++ b/build-parent/pom.xml
@@ -295,8 +295,10 @@
                                             <!-- Ban checker-qual, we don't use Checker Framework -->
                                             <exclude>org.checkerframework:checker-qual</exclude>
                                             <exclude>com.google.code.findbugs:jsr305</exclude>
+                                            <!-- We use our own impl here, including this one causes problems-->
+                                            <exclude>org.jboss.resteasy:resteasy-context-propagation</exclude>
                                         </excludes>
-                                        <includes>
+                                        <includes>a
                                             <!-- this is for REST Assured -->
                                             <include>jakarta.xml.bind:jakarta.xml.bind-api:*:*:test</include>
                                         </includes>

--- a/extensions/mutiny/runtime/pom.xml
+++ b/extensions/mutiny/runtime/pom.xml
@@ -26,12 +26,6 @@
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-smallrye-context-propagation</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>org.jboss.resteasy</groupId>
-                    <artifactId>resteasy-context-propagation</artifactId>
-                </exclusion>
-            </exclusions>
         </dependency>
         <dependency>
             <groupId>io.smallrye.reactive</groupId>

--- a/extensions/resteasy-common/runtime/pom.xml
+++ b/extensions/resteasy-common/runtime/pom.xml
@@ -26,7 +26,6 @@
             <groupId>org.jboss.resteasy</groupId>
             <artifactId>resteasy-core</artifactId>
         </dependency>
-        <!--
         <dependency>
             <groupId>org.jboss.resteasy</groupId>
             <artifactId>resteasy-context-propagation</artifactId>
@@ -37,7 +36,6 @@
                 </exclusion>
             </exclusions>
         </dependency>
-        -->
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-arc</artifactId>

--- a/extensions/resteasy-common/runtime/pom.xml
+++ b/extensions/resteasy-common/runtime/pom.xml
@@ -57,6 +57,7 @@
                         <excludedArtifact>javax.activation:javax.activation-api</excludedArtifact>
                         <excludedArtifact>javax.activation:activation</excludedArtifact>
                         <excludedArtifact>jakarta.ws.rs:jakarta.ws.rs-api</excludedArtifact>
+                        <excludedArtifact>org.jboss.resteasy:resteasy-context-propagation</excludedArtifact>
                     </excludedArtifacts>
                 </configuration>
             </plugin>

--- a/extensions/resteasy-common/runtime/pom.xml
+++ b/extensions/resteasy-common/runtime/pom.xml
@@ -27,16 +27,6 @@
             <artifactId>resteasy-core</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.jboss.resteasy</groupId>
-            <artifactId>resteasy-context-propagation</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>io.smallrye</groupId>
-                    <artifactId>smallrye-context-propagation</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-arc</artifactId>
         </dependency>

--- a/extensions/resteasy-common/runtime/src/main/java/io/quarkus/resteasy/common/runtime/ResteasyContextProvider.java
+++ b/extensions/resteasy-common/runtime/src/main/java/io/quarkus/resteasy/common/runtime/ResteasyContextProvider.java
@@ -1,0 +1,40 @@
+package io.quarkus.resteasy.common.runtime;
+
+import java.util.Collections;
+import java.util.Map;
+
+import org.eclipse.microprofile.context.spi.ThreadContextProvider;
+import org.eclipse.microprofile.context.spi.ThreadContextSnapshot;
+import org.jboss.resteasy.core.ResteasyContext;
+
+public class ResteasyContextProvider implements ThreadContextProvider {
+
+    private static final String JAXRS_CONTEXT = "JAX-RS";
+
+    @Override
+    public ThreadContextSnapshot currentContext(Map<String, String> props) {
+        Map<Class<?>, Object> context = ResteasyContext.getContextDataMap();
+        return () -> {
+            ResteasyContext.pushContextDataMap(context);
+            return () -> {
+                ResteasyContext.removeContextDataLevel();
+            };
+        };
+    }
+
+    @Override
+    public ThreadContextSnapshot clearedContext(Map<String, String> props) {
+        Map<Class<?>, Object> context = Collections.emptyMap();
+        return () -> {
+            ResteasyContext.pushContextDataMap(context);
+            return () -> {
+                ResteasyContext.removeContextDataLevel();
+            };
+        };
+    }
+
+    @Override
+    public String getThreadContextType() {
+        return JAXRS_CONTEXT;
+    }
+}

--- a/extensions/resteasy-common/runtime/src/main/resources/META-INF/services/org.eclipse.microprofile.context.spi.ThreadContextProvider
+++ b/extensions/resteasy-common/runtime/src/main/resources/META-INF/services/org.eclipse.microprofile.context.spi.ThreadContextProvider
@@ -1,0 +1,1 @@
+io.quarkus.resteasy.common.runtime.ResteasyContextProvider

--- a/extensions/smallrye-context-propagation/deployment/src/main/java/io/quarkus/smallrye/context/deployment/SmallRyeContextPropagationProcessor.java
+++ b/extensions/smallrye-context-propagation/deployment/src/main/java/io/quarkus/smallrye/context/deployment/SmallRyeContextPropagationProcessor.java
@@ -40,14 +40,6 @@ class SmallRyeContextPropagationProcessor {
         List<ContextManagerExtension> discoveredExtensions = new ArrayList<>();
         for (Class<?> provider : ServiceUtil.classesNamedIn(Thread.currentThread().getContextClassLoader(),
                 "META-INF/services/" + ThreadContextProvider.class.getName())) {
-            if (provider.getName().equals("org.jboss.resteasy.context.ResteasyContextProvider")) {
-                try {
-                    Class.forName("org.jboss.resteasy.core.ResteasyContext", false,
-                            Thread.currentThread().getContextClassLoader());
-                } catch (ClassNotFoundException e) {
-                    continue; // resteasy is not being used so ditch this context provider
-                }
-            }
             try {
                 discoveredProviders.add((ThreadContextProvider) provider.newInstance());
             } catch (InstantiationException | IllegalAccessException e) {

--- a/extensions/smallrye-context-propagation/runtime/pom.xml
+++ b/extensions/smallrye-context-propagation/runtime/pom.xml
@@ -29,16 +29,6 @@
             <artifactId>jakarta.enterprise.cdi-api</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.jboss.resteasy</groupId>
-            <artifactId>resteasy-context-propagation</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>org.jboss.resteasy</groupId>
-                    <artifactId>resteasy-core</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-core</artifactId>
         </dependency>

